File: 4-2.rb
#!/usr/bin/env ruby

require 'set'

class Passport
  @@fields = [:byr, :iyr, :eyr, :hgt, :hcl, :ecl, :pid, :cid]
  @@fields.each do |field|
    attr_accessor field
  end

  @@eye_colours = Set["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]

  def initialize(str)
    str.split.each do |fv|
      field, value = fv.split(':')
      self.send(field + "=", value)
    end
  end

  def valid_byr?
    @byr.to_i.between?(1920, 2002)
  end

  def valid_iyr?
    @iyr.to_i.between?(2010, 2020)
  end

  def valid_eyr?
    @eyr.to_i.between?(2020, 2030)
  end

  def valid_hgt?
    return false unless @hgt
    case @hgt[-2 .. -1]
    when 'cm'
      return @hgt[0..-3].to_i.between?(150, 193)
    when 'in'
      return @hgt[0..-3].to_i.between?(59, 76)
    end
    false
  end

  def valid_hcl?
    @hcl =~ /#[[:xdigit:]]{6}/
  end

  def valid_ecl?
    @@eye_colours.member? @ecl
  end

  def valid_pid?
    @pid =~ /^[[:digit:]]{9}$/
  end

  def valid_cid?
    true
  end

  def valid?
    @@fields.all? {|f| self.send ('valid_' + f.to_s + '?').to_sym}
  end
end

valid = 0
File.read('4.input').split("\n\n").each do |s|
  valid += 1 if Passport.new(s).valid?
end

print("#{valid}\n")

File: 4-2.py
#!/usr/bin/env python

import re

class Passport:
  fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid", "cid"]
  eye_colours = {"amb", "blu", "brn", "gry", "grn", "hzl", "oth"}

  def __init__(self, str):
    for fv in str.split():
      field, value = fv.split(':')
      setattr(self, field, value)

  def valid(self):
    return all(map(lambda f: getattr(self, "valid_" + f, None)(),
                   self.fields))

  def valid_byr(self):
    return 1920 <= int(getattr(self, "byr", 0)) <= 2002

  def valid_iyr(self):
    return 2010 <= int(getattr(self, "iyr", 0)) <= 2020

  def valid_eyr(self):
    return 2020 <= int(getattr(self, "eyr", 0)) <= 2030

  def valid_hgt(self):
    if not getattr(self, "hgt", None):
      return False
    if self.hgt[-2:] == 'cm':
      return 150 <= int(self.hgt[0:-2]) <= 193
    elif self.hgt[-2:] == 'in':
      return 59 <= int(self.hgt[0:-2]) <= 76
    return False

  def valid_hcl(self):
    return re.fullmatch('#[0-9A-Fa-f]{6}', getattr(self, "hcl", ""))

  def valid_ecl(self):
    return getattr(self, "ecl", None) in self.eye_colours

  def valid_pid(self):
    return re.fullmatch('[0-9]{9}', getattr(self, "pid", ""))

  def valid_cid(self):
    return True

valid = 0
for s in open('4.input').read().split("\n\n"):
  if Passport(s).valid():
    valid += 1

print(valid)
